### DISCUSSION TITLE
'random' module incomplete ?

### DISCUSSION THREAD

**Discussion by gbroulet** [micropython/micropython#11578] [ESP8266] [answered]
Hello,

I just uploaded the latest release for ESP8266 2M+, V1.20. The board is a ESP8266-12F NodeMCU Lolin V3.

According to the docs, the "random" module should come with a heap of functions (like randint() or uniform()). However, when trying to access these on my 8266 dev board, it appears only seed() and getrandbits() are present.

```py
>>> import random
>>> random.uniform(13, 25)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AttributeError: 'module' object has no attribute 'uniform'
```

Could someone please tell me what I am missing ?

Thanks in advance for your help !

EDIT : I also had a question about espnow, but found my answer, so deleted it

**Comment by robert-hh** [comment, marked as answer]
As you tell, the esp8266 random module does implement only a small subset of the random module to save flash space for devices with small memory. Building your own firmware, you can change that by setting:

`#define MICROPY_PY_URANDOM_EXTRA_FUNCS          (1)`

In the mpconfigport.h of the ESP8266 port, or better in the mpconfigboard.h of the GENERIC board section. Generally, the configuration could be changed to support more features with the 2MB version of the firmware. Only someone has to make the change, test it and make a PR for it.

**Reply by gbroulet** [reply]
Thank you very much, I will look into it

**Reply by jimmo** [reply]
@gbroulet I know you're looking for `random.uniform` but some of the other functions missing when `MICROPY_PY_URANDOM_EXTRA_FUNCS` isn't enabled (e.g. `random.randint`, `random.randrange`), plus others from CPython such as  `random.shuffle`, can be added using the random module from micropython-lib (which can be installed via `mip`).

We might want to consider adding `random.random`, `random.choice` and `random.uniform` to the micropython-lib version.

**Comment by konomikitten** [comment]
I know this has been answered but for future people like myself wanting randrange:

```py
import random

def randrange(start, stop):
    return random.getrandbits(16) % (stop + 1 - start) + start
```

**Reply by robert-hh** [reply]
That works only well if the span of the range is a power of 2. Otherwise, the numbers from the ĺower end of the span are more frequent. It looks better if the span is much smaller than 2 ** number of bits for getrandbits.

**Reply by rsharris** [reply]
As @robert-hh indicates, unless the range is small (much smaller than 2**16 in this case) the result of @konomikitten 's function will introduce bias. I learned this lesson the hard way when using 32 bit randoms to choose a location in the human genome. My range was the length of the human genome, roughly 3G. And so the first 1.2G positions were twice as likely to be generated than the remaining 1.8G.

This 2019 article describes an algorithm to generate unbiased uniform variates: Lemire, Daniel. "Fast random integer generation in an interval." ACM Transactions on Modeling and Computer Simulation (TOMACS) 29.1 (2019): 1-12.

**Reply by konomikitten** [reply]
Is there a better solution?

**Reply by scruss** [reply]
> Is there a better solution?

Yes. As jimmo said, install [random](https://github.com/micropython/micropython-lib/tree/master/python-stdlib/random) from micropython-lib.

**Reply by rsharris** [reply]
`Yes. As jimmo said, install [random] from micropython-lib.`

Well, that's only a better solution if you're in an environment where you can do that. In my case, I'm using pyscript and (to the best of my knowledge) my only choice is whatever micropython pyscript.net provides. That micropython build *does* include randrange() and uniform(). But it is missing Random() which would allow me to have different PRNG states for different parts of my process. 

To be clear, I'm not complaining. I'm using the micropython variant of pyscript because I want to have a lean footprint. I expect/anticipate/encounter some functional shortcomings and work around them.

The Lemire algorithm (it's algorithm 5 in that paper) is fast, unbiased, easy to implement, and not too hard to test. If I needed randrange() and it wasn't available, that's what I'd use.

**Reply by Josverl** [reply]
@rsharris , in pyscript you can use external modules using MIP or url

From pyscript docs, FAQ

In MicroPython, the package you want to use has not been ported into the micropython-lib package repository. If you want to use a pure Python package with MicroPython, use the files configuration option to manually copy the package onto the file system, or use a URL to reference the package.

https://docs.pyscript.net/2025.3.1/faq/

**Reply by rsharris** [reply]
@Josverl Thanks, that might be the workaround I need. I'll have to explore that.

I did read the FAQ (recently, even) but my memory did not make a connection back to that when I encountered the missing Random(). And it's not exactly the same situation — the package *has* been ported but it lacks some functionality it has in 'normal' python.

Thanks for pointing me at that.

**Reply by Josverl** [reply]
When I compare the stubs I created on pyscript/webassembly (v1.24.1 at the time) and for instance rp2 I seen no difference in methods and functions between the two, and both include `randrange()`. 

Do you perhaps have a file named `random.py` in your project ? that would throw off the import.

**Comment by rsharris** [comment]
Interesting. No, I don't have a random.py in my project.

I'm using `https://pyscript.net/releases/2025.3.1/core.js` with `script type="mpy"`. When I tried `from random import Random` in my .py file, it threw an ImportError exception (module existed, but Random didn't). If I switched to `script type="py"` that exception disappeared.

To discover what was available in random, in the mpy case, I did `import random` and then printed `random.__dict__` to the console. What exists is __name__, __init__, getrandbits, seed, randrange, randint, choice, random, and uniform. All these were listed as functions except __name__. I didn't try the same with the py case, but if I look at `random.__dict__` in my desktop python3 shell, random.Random is present as a class.

**Reply by rsharris** [reply]
I've now also looked at `random.__dict__` for the `script type="py"` case. It is very similar to what I get in my desktop python3 shell, and it as Random as a class.
